aspnetcore/blazor/webassembly-lazy-load-assemblies.md

```diff
@@ -0,0 +1,163 @@
+
+```xml
+<ItemGroup>
+  <BlazorWebAssemblyLazyLoad Include="GrantImaharaRobotControls.dll" />
+</ItemGroup>
+```
+
+Only assemblies that are used by the app can be lazily loaded. The linker strips unused assemblies from published output.
+
+## `Router` component
+
+Code in the `Router` component determines when the assemblies marked for lazy loading are loaded.
+
+In the app's `Router` component (`App.razor`):
+
+* Add an `OnNavigateAsync` callback. The `OnNavigateAsync` handler is invoked when the user visits a route for the first time or navigates to a new route.
+* Add a [List](xref:System.Collections.Generic.List%601)\<<xref:System.Reflection.Assembly>> (for example, named `lazyLoadedAssemblies`) to the component. The assemblies are passed back to the <xref:Microsoft.AspNetCore.Components.Routing.Router.AdditionalAssemblies> collection in case the assemblies contain routable components. The framework searches the assemblies for routes and updates the route collection if any new routes are found.
+
+```razor
+@using System.Reflection
+
+<Router AppAssembly="@typeof(Program).Assembly" 
+    AdditionalAssemblies="@lazyLoadedAssemblies" OnNavigateAsync="@OnNavigateAsync">
+    ...
+</Router>
+
+@code {
+    private List<Assembly> lazyLoadedAssemblies = new List<Assembly>();
+
+    private async Task OnNavigateAsync(NavigationContext args)
+    {
+    }
+}
+```
+
+### Assembly load logic in `OnNavigateAsync`
+
+`OnNavigateAsync` has a `NavigationContext` parameter that provides information about the current asynchronous navigation event, including the target path (`Path`) and the cancellation token (`CancellationToken`):
+
+* The `Path` property is the user's destination path relative to the app's base path, such as `/robot`.
+* The `CancellationToken` can be used to observe the cancellation of the asynchronous task. `OnNavigateAsync` automatically cancels the currently running navigation task when the user navigates away from a page.
+
+Inside `OnNavigateAsync`, implement logic to determine the assemblies to load. Options include:
+
+* Conditional checks inside the `OnNavigateAsync` method.
+* A lookup table that maps routes to assembly names, either injected into the component or implemented within the [`@code`](xref:mvc/views/razor#code) block.
+
+`LazyAssemblyLoader` is a framework-provided singleton service for loading assemblies. Inject `LazyAssemblyLoader` into the `Router` component:
+
+```razor
+...
+@using Microsoft.AspNetCore.Components.WebAssembly.Services
+@inject LazyAssemblyLoader assemblyLoader
+
+...
+```
+
+The `LazyAssemblyLoader` provides the `LoadAssembliesAsync` method that:
```

Comment:
Let's add a note here that the `LazyAssemblyLoader` is compatible with pre-rendering scenarios. When pre-rendering, we don't lazy-load since all of the assemblies are on the server where the pre-rendering is done and the JSRuntime is not available when pre-rendering.

Reply:
Covered on Line 87

Reply:
> "the JSRuntime is not available" ... seemed less important given that the app has access to assemblies directly that are already loaded, and that's already stated.

That's a good point. This is probably not an important detail for an end-user to know so we can nuke it.